File: Stack/stack.py
# 3.6.2

class Stack(object):

    def __init__(self, length):
        self._items = [None] * length
        self._index = 0

    def __len__(self):
        return self._index

    def is_empty(self):
        return self._index == 0

    def is_full(self):
        return self._index == len(self._items)

    def peek(self):

        if self.is_empty():
            raise IndexError("Cannot peek at an empty stack!")

        return self._items[self._index - 1]

    def pop(self):

        if self.is_empty():
            raise IndexError("Cannot pop from an empty stack!")

        self._index -= 1
        item = self._items[self._index]
        self._items[self._index] = None
        return item

    def push(self, item):

        if self.is_full():
            raise IndexError("Cannot push into a full stack!")

        self._items[self._index] = item
        self._index += 1

File: Stack/stack.kt
// 1.1.3

internal class Stack<T>(size: Int) {

  internal val isEmpty: Boolean
    get() = this.index == 0

  internal val isFull: Boolean
    get() = this.index == this.items.size

  internal val size: Int
    get() = this.index

  private var items: Array<Any?>
  private var index: Int

  internal init {
    this.items = arrayOfNulls<Any?>(size)
    this.index = 0
  }

  internal fun peek(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot peek at an empty stack!")
    }

    return this.items[this.index - 1] as T?
  }

  internal fun pop(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot pop from an empty stack!")
    }

    this.index -= 1
    val item = this.items[this.index]
    this.items[this.index] = null
    return item as T?
  }

  internal fun push(item: T?): Unit {

    if (this.isFull) {
      throw Exception("Cannot push into a full stack!")
    }

    this.items[this.index] = item
    this.index += 1
  }

}
